File: aoc-2015/src/main/java/com/capital7software/aoc/aoc2015/days/Day12.java
package com.capital7software.aoc.aoc2015.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.string.JsonFun;
import java.time.Instant;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * --- Day 12: JSAbacusFramework.io ---<br><br>
 * Santa's Accounting-Elves need help balancing the books after a recent order. Unfortunately,
 * their accounting software uses a peculiar storage format. That's where you come in.
 *
 * <p><br>
 * They have a JSON document which contains a variety of things: arrays ([1,2,3]),
 * objects ({"a":1, "b":2}), numbers, and strings. Your first job is to simply
 * find all the numbers throughout the document and add them together.
 *
 * <p><br>
 * For example:
 *
 * <p><br>
 * <code>
 * [1,2,3] and {"a":2,"b":4} both have a sum of 6.<br>
 * [[[3]]] and {"a":{"b":4},"c":-1} both have a sum of 3.<br>
 * {"a":[-1,1]} and [-1,{"a":1}] both have a sum of 0.<br>
 * [] and {} both have a sum of 0.<br>
 * </code>
 *
 * <p><br>
 * You will not encounter any strings containing numbers.
 *
 * <p><br>
 * What is the sum of all numbers in the document?
 *
 * <p><br>
 * Your puzzle answer was 156366.
 *
 * <p><br>
 * --- Part Two ---<br><br>
 * Uh oh - the Accounting-Elves have realized that they double-counted everything red.
 *
 * <p><br>
 * Ignore any object (and all of its children) which has any property with the
 * value "red". Do this only for objects ({...}), not arrays ([...]).
 *
 * <p><br>
 * <code>
 * [1,2,3] still has a sum of 6.<br>
 * [1,{"c":"red","b":2},3] now has a sum of 4, because the middle object is ignored.<br>
 * {"d":"red","e":[1,2,3,4],"f":5} now has a sum of 0, because the entire structure is ignored.<br>
 * [1,"red",5] has a sum of 6, because "red" in an array has no effect.<br>
 * </code>
 *
 * <p><br>
 * Your puzzle answer was 96852.
 */
public class Day12 implements AdventOfCodeSolution {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day12.class);

  /**
   * Instantiates the solution instance.
   */
  public Day12() {

  }

  @Override
  public String getDefaultInputFilename() {
    return "inputs/input_day_12-01.txt";
  }

  @Override
  public void runPart1(List<String> input) {
    for (var line : input) {
      var start = Instant.now();
      var sum = sumNumbersInString(line);
      var end = Instant.now();
      LOGGER.info("The sum of all numbers in the JSON string is: {}", sum);
      logTimings(LOGGER, start, end);
    }
  }

  @Override
  public void runPart2(List<String> input) {
    for (var line : input) {
      var start = Instant.now();
      var sum = sumNumbersInStringSkippingRedObjects(line);
      var end = Instant.now();
      LOGGER.info("The sum of all numbers in the JSON string skipping red objects is: {}", sum);
      logTimings(LOGGER, start, end);
    }
  }

  /**
   * Returns the sum of all the numbers encountered in the specified JSON String.
   *
   * @param line The JSON String to sum the numbers from.
   * @return The sum of all the numbers encountered in the specified JSON String.
   */
  public long sumNumbersInString(String line) {
    return JsonFun.sumOfAllNumbers(line);
  }


  /**
   * Returns the sum of all the numbers encountered in the specified JSON String
   * that are not in a Red object or sub-object.
   *
   * @param line The JSON String to sum the numbers from.
   * @return The sum of all the numbers encountered in the specified JSON String
   *     that are not in a Red object or sub-object.
   */
  public long sumNumbersInStringSkippingRedObjects(String line) {
    return JsonFun.sumOfAllNumbersSkippingObjectsWithRedPropertyValues(line);
  }
}

File: aoc-2016/src/main/kotlin/com/capital7software/aoc/aoc2016aoc/days/Day12.kt
package com.capital7software.aoc.aoc2016aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.computer.SmallComputer
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * --- Day 12: Leonardo's Monorail ---
 * You finally reach the top floor of this building: a garden with a slanted glass ceiling.
 * Looks like there are no more stars to be had.
 *
 * While sitting on a nearby bench amidst some tiger lilies, you manage to decrypt some of the
 * files you extracted from the servers downstairs.
 *
 * According to these documents, Easter Bunny HQ isn't just this building - it's a
 * collection of buildings in the nearby area. They're all connected by a local monorail,
 * and there's another building not far from here! Unfortunately, being night, the
 * monorail is currently not operating.
 *
 * You remotely connect to the monorail control systems and discover that the boot
 * sequence expects a password. The password-checking logic (your puzzle input) is
 * easy to extract, but the code it uses is strange: it's assembunny code designed
 * for the new computer you just assembled. You'll have to execute the code and get
 * the password.
 *
 * The assembunny code you've extracted operates on four registers (a, b, c, and d)
 * that start at 0 and can hold any integer. However, it seems to make use of only
 * a few instructions:
 *
 * - cpy x y **copies** x (either an integer or the value of a register) into register y.
 * - inc x **increases** the value of register x by one.
 * - dec x **decreases** the value of register x by one.
 * - jnz x y **jumps** to an instruction y away (positive means forward; negative means
 * backward), but only if x is not zero.
 * - The jnz instruction moves relative to itself: an offset of -1 would continue at
 * the previous instruction, while an offset of 2 would **skip over** the next instruction.
 *
 * For example:
 * ```
 * cpy 41 a
 * inc a
 * inc a
 * dec a
 * jnz a 2
 * dec a
 * ```
 * The above code would set register a to 41, increase its value by 2, decrease its
 * value by 1, and then skip the last dec a (because a is not zero, so the jnz a 2
 * skips it), leaving register a at 42. When you move past the last instruction,
 * the program halts.
 *
 * After executing the assembunny code in your puzzle input, **what value is left
 * in register a?**
 *
 * --- Part Two ---
 * As you head down the fire escape to the monorail, you notice it didn't start;
 * register c needs to be initialized to the position of the ignition key.
 *
 * If you instead initialize register c to be 1, what value is now left in register a?
 */
class Day12 : AdventOfCodeSolution {
  companion object {
    private val log: Logger = LoggerFactory.getLogger(Day12::class.java)
  }

  override fun getDefaultInputFilename(): String {
    return "inputs/input_day_12-01.txt"
  }

  override fun runPart1(input: List<String>) {
    val register = "a"
    val start = Instant.now()
    val answer = getValueInRegister(input, register)
    val end = Instant.now()

    log.info("$answer is the value in register $register after running the program!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: List<String>) {
    val register = "a"
    val destRegister = "c"
    val destValue = 1L
    val start = Instant.now()
    val answer = setAndGetValueInRegister(input, register, destRegister, destValue)
    val end = Instant.now()

    log.info("$answer is the value in register $register after running the program!")
    logTimings(log, start, end)
  }

  /**
   * Runs the specified program and returns the value of the specified register after the
   * program has run.
   *
   * @param input The [List] of [String] instructions to parse and execute.
   * @return The value of the specified register after the program has run.
   */
  @SuppressFBWarnings
  fun getValueInRegister(input: List<String>, register: String): Long {
    val computer = SmallComputer.buildSmallComputer(input)
    computer.run()
    return computer[register]
  }

  /**
   * Runs the specified program and returns the value of the specified register after the
   * program has run.
   *
   * @param input The [List] of [String] instructions to parse and execute.
   * @return The value of the specified register after the program has run.
   */
  @SuppressFBWarnings
  fun setAndGetValueInRegister(
      input: List<String>,
      register: String,
      destRegister: String,
      destValue: Long
  ): Long {
    val computer = SmallComputer.buildSmallComputer(input)
    computer[destRegister] = destValue
    computer.run()
    return computer[register]
  }
}
